internal/mixpanel/client.go

```diff
@@ -29,13 +36,20 @@ func NewClient(serviceAccountUsername, serviceAccountSecret *string) (*Client, e
 
 	c.AuthHeader = "Basic " + *serviceAccountUsername + ":" + *serviceAccountSecret
 
+	c.Semaphore = semaphore.NewWeighted(concurrentRequests)
+
 	return &c, nil
 }
 
 func (c *Client) doRequest(req *http.Request) ([]byte, error) {
 	req.Header.Add("Authorization", c.AuthHeader)
 
+	err := c.Semaphore.Acquire(req.Context(), 1)
+	if err != nil {
+		return nil, err
```

Comment:
Shouldn't you release the semaphore before returning?

Reply:
I'm unsure how this is handled because there will be an error if Semaphore is not Acquired so we don't need to release it?

I did made sure to release the semaphore before the error handling of the request.

Reply:
OK so the error can only happen if it's not acquired at all? There can't be several types of error?

Reply:
Yes the call is blocking until it can be acquired so it won't return an error if the semaphore is already occupied, the package doesn't explicitly say what kind of errors can happen but I don't expect any error here. I only added the error checking because of the complaint from the linter.

https://pkg.go.dev/golang.org/x/sync/semaphore#Weighted.Acquire